src/capstone.rs

```diff
@@ -1,63 +1,146 @@
+
+/// Represents an empty set of `ExtraMode`.
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub struct EmptyExtraModeIter(PhantomData<()>);
+
+impl Iterator for EmptyExtraModeIter {
+    type Item = ExtraMode;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        None
+    }
+}
+
 impl Capstone {
-    /// Create a new instance of the decompiler.
+    /// Create a new instance of the decompiler using the builder pattern interface.
+    /// This is the recommended interface to `Capstone`.
+    ///
+    /// ```
+    /// use capstone::{Capstone, BuildsCapstone, x86};
+    /// let cs = Capstone::new().x86().mode(x86::ArchMode::Mode32).build();
+    /// ```
+    pub fn new() -> CapstoneBuilder {
+        CapstoneBuilder::new()
+    }
+
+    /// Create a new instance of the decompiler using the "raw" interface.
+    /// The user must ensure that only sensical `Arch`/`Mode` combinations are used.
     ///
     /// ```
-    /// use capstone::{Arch, Capstone, Mode};
-    /// let cs = Capstone::new(Arch::X86, Mode::Mode64);
+    /// use capstone::{Arch, Capstone, NO_EXTRA_MODE, Mode};
+    /// let cs = Capstone::new_raw(Arch::X86, Mode::Mode64, NO_EXTRA_MODE, None);
     /// assert!(cs.is_ok());
     /// ```
-    pub fn new(arch: Arch, mode: Mode) -> CsResult<Capstone> {
+    pub fn new_raw<T: Iterator<Item = ExtraMode>>(
+        arch: Arch,
+        mode: Mode,
+        extra_mode: T,
+        endian: Option<Endian>,
+    ) -> CsResult<Capstone> {
         let mut handle = 0;
         let csarch: cs_arch = arch.into();
         let csmode: cs_mode = mode.into();
-        let err = unsafe { cs_open(csarch, csmode, &mut handle) };
 
-        if cs_err::CS_ERR_OK == err {
-            let mut opt_state: HashMap<cs_opt_type, libc::size_t> = HashMap::new();
-            opt_state.insert(cs_opt_type::CS_OPT_SYNTAX,
-                             CS_OPT_SYNTAX_DEFAULT as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_DETAIL,
-                             cs_opt_value::CS_OPT_OFF as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_MODE, mode as libc::size_t);
-            opt_state.insert(cs_opt_type::CS_OPT_MEM, 0);
-            opt_state.insert(cs_opt_type::CS_OPT_SKIPDATA,
-                             cs_opt_value::CS_OPT_OFF as libc::size_t);
-
-            Ok(Capstone {
-                   csh: handle,
-                   cs_option_state: opt_state,
-                   _arch: arch,
-               })
-        } else {
-            Err(err.into())
-        }
-    }
+        let endian = match endian {
+            Some(endian) => cs_mode::from(endian) as usize,
+            None => 0,
+        };
+        let extra_mode = Self::extra_mode_value(extra_mode);
+
+        let err = unsafe {
+            let combined_mode: usize = csmode as usize | endian | extra_mode;
+            let combined_mode: cs_mode = mem::transmute(combined_mode as CsModeRepr);
```

Comment:
Is transmute  necessary?

Reply:
Yes, because Rust does not let us create an enum with a non-existent discriminant. In order to mitigate the risk of transmuting to the wrong size/sign, I defined the `CsModeRepr` type. There's a test to ensure that the size is the same in `src/constants.rs`.